File: day16/Main.hs
import           Data.List
import           Data.Bifunctor
import           Data.Map                       ( Map )
import qualified Data.Map                      as Map
import           Text.Parsec
import           Text.Parsec.String             ( Parser )

type Ticket = [Int]
newtype Constraint = Constraint [(Int, Int)]
  deriving (Show)
type NamedConstraints = Map String Constraint

parseUnsigned :: Parser Int
parseUnsigned = read <$> many1 digit

parseBounds :: Parser (Int, Int)
parseBounds = do
  lo <- parseUnsigned
  char '-'
  hi <- parseUnsigned
  return (lo, hi)

parseConstraint :: Parser (String, Constraint)
parseConstraint = do
  name <- many1 (lower <|> char ' ')
  string ": "
  bounds <- parseBounds `sepBy` string " or "
  newline
  return (name, Constraint bounds)

parseTicket :: Parser Ticket
parseTicket = do
  vals <- parseUnsigned `sepBy` char ','
  newline
  return vals

parseInput :: Parser (NamedConstraints, Ticket, [Ticket])
parseInput = do
  constraints <- many1 parseConstraint
  newline
  string "your ticket:"
  newline
  yourTicket <- parseTicket
  newline
  string "nearby tickets:"
  newline
  nearbyTickets <- many1 parseTicket
  return (Map.fromList constraints, yourTicket, nearbyTickets)

meetsConstraint :: Int -> Constraint -> Bool
meetsConstraint v (Constraint bounds) =
  any (\(lo, hi) -> lo <= v && v <= hi) bounds

constraintsMet :: Int -> NamedConstraints -> [Constraint]
constraintsMet val constraints =
  filter (meetsConstraint val) (Map.elems constraints)

scanningErrorRate :: NamedConstraints -> [Ticket] -> Int
scanningErrorRate constraints =
  sum . map (sum . filter (\e -> null (constraintsMet e constraints)))

validTickets :: NamedConstraints -> [Ticket] -> [Ticket]
validTickets constraints =
  filter (not . any (\e -> null (constraintsMet e constraints)))

-- Rotates tickets from a form like
--      C1 C2 C3 C4
--   T1 A  B  C  D
--   T2 E  F  G  H
-- to
--      T1 T2
--   C1 A  E
--   C2 B  F
--   C3 C  G
--   C4 D  H
rotateTickets :: [Ticket] -> [[Int]]
rotateTickets []       = []
rotateTickets ([] : _) = []
rotateTickets l        = map head l : rotateTickets (map tail l)

mfind a mp = case Map.lookup a mp of
  Just b -> b

-- Returns a list of constraints that are valid for all values in a ticket column.
validForColumn :: NamedConstraints -> [Int] -> [String]
validForColumn constraints values = Map.keys $ Map.filter
  (\constraint -> all (`meetsConstraint` constraint) values)
  constraints

-- For each column of ticket values, return a list of valid constraints for that column.
columnOptions :: NamedConstraints -> [[Int]] -> [[String]]
columnOptions constraints = map (validForColumn constraints)

-- Given a list of valid constraints for each ticket column, finds a permutation
-- of exactly one constraint per column satisfying each column's constraint options.
solve :: [[String]] -> [String]
solve options =
  let withColumn = zip [0 ..] options
      sortByConstraints opts = sortOn (length . snd) opts
      solve' :: [(Int, [String])] -> [(Int, String)]
      solve' [] = []
      solve' ((id, [solved]) : rest) =
          let pruneSolved = map (second (filter (/= solved))) rest
          in  (id, solved) : solve' (sortByConstraints pruneSolved)
      solved = solve' (sortByConstraints withColumn)
  in  map snd (sortOn fst solved)

main = do
  input <- readFile "input.txt"
  let (constraints, yourTicket, nearbyTickets) =
        case parse parseInput "" input of
          Left  e -> error (show e)
          Right r -> r
  print "Part 1:"
  print $ scanningErrorRate constraints nearbyTickets
  let ticketValues =
        rotateTickets (yourTicket : validTickets constraints nearbyTickets)
  print "Part 2:"
  let constraintOrder = solve (columnOptions constraints ticketValues)
  print $ product
    (map
      snd
      (filter (isPrefixOf "departure" . fst) (zip constraintOrder yourTicket))
    )

File: day16/main.go
package main

import (
	"fmt"
	"io/ioutil"
	"regexp"
	"strconv"
	"strings"
)

func check(e error) {
	if e != nil {
		panic(e)
	}
}

type bound struct {
	lo, hi int
}

func (b bound) validate(v int) bool {
	return b.lo <= v && v <= b.hi
}

type constraint struct {
	bounds []bound
}

func (c constraint) validate(v int) bool {
	for _, bound := range c.bounds {
		if bound.validate(v) {
			return true
		}
	}
	return false
}

type ticket struct {
	values []int
}

// Returns the scanning error rate for a ticket given a set of constraints, and
// whether the ticket is valid (has all fields meeting the constraints).
func (t ticket) scanErrorRate(constraints []constraint) (int, bool) {
	errorRate := 0
	valid := true
checkValues:
	for _, v := range t.values {
		for _, c := range constraints {
			if c.validate(v) {
				continue checkValues
			}
		}
		errorRate += v
		valid = false
	}
	return errorRate, valid
}

func (t ticket) col(i int) int {
	return t.values[i]
}

type data struct {
	fields        map[string]constraint
	yourTicket    ticket
	nearbyTickets []ticket
}

func parse(input string) data {
	splits := strings.Split(input, "\n\n")

	fields := map[string]constraint{}
	reField := regexp.MustCompile(`(.+): (\d+)-(\d+) or (\d+)-(\d+)`)
	for _, sfield := range strings.Split(splits[0], "\n") {
		match := reField.FindStringSubmatch(sfield)
		lo1, _ := strconv.Atoi(match[2])
		hi1, _ := strconv.Atoi(match[3])
		lo2, _ := strconv.Atoi(match[4])
		hi2, _ := strconv.Atoi(match[5])
		bounds := []bound{{lo1, hi1}, {lo2, hi2}}
		fields[match[1]] = constraint{bounds}
	}

	parseTicket := func(s string) ticket {
		values := []int{}
		for _, sv := range strings.Split(s, ",") {
			v, _ := strconv.Atoi(sv)
			values = append(values, v)
		}
		return ticket{values}
	}

	yourTicket := parseTicket(strings.Split(splits[1], "\n")[1])
	snearbyTickets := strings.Split(splits[2], "\n")
	nearbyTickets := []ticket{}
	for _, ticket := range snearbyTickets[1 : len(snearbyTickets)-1] {
		nearbyTickets = append(nearbyTickets, parseTicket(ticket))
	}

	return data{fields, yourTicket, nearbyTickets}
}

func (d data) values() (map[string]constraint, ticket, []ticket) {
	return d.fields, d.yourTicket, d.nearbyTickets
}

func listConstraints(fields map[string]constraint) []constraint {
	lst := []constraint{}
	for _, c := range fields {
		lst = append(lst, c)
	}
	return lst
}

func listFieldNames(fields map[string]constraint) []string {
	lst := []string{}
	for n := range fields {
		lst = append(lst, n)
	}
	return lst
}

func nearbyTicketErrorRate(fields map[string]constraint, tickets []ticket) int {
	rate := 0
	constraints := listConstraints(fields)
	for _, ticket := range tickets {
		prate, _ := ticket.scanErrorRate(constraints)
		rate += prate
	}
	return rate
}

func filterValid(fields map[string]constraint, tickets []ticket) []ticket {
	valid := []ticket{}
	constraints := listConstraints(fields)
	for _, ticket := range tickets {
		if _, ok := ticket.scanErrorRate(constraints); ok {
			valid = append(valid, ticket)
		}
	}
	return valid
}

func lst2set(lst []string) map[string]bool {
	set := map[string]bool{}
	for _, s := range lst {
		set[s] = true
	}
	return set
}

func singleton(m map[string]bool) string {
	for s := range m {
		return s
	}
	panic("Not a singleton")
}

func solveFieldOrder(fields map[string]constraint, tickets []ticket) []string {
	// Part 1 :: assign each column a set of possible field names.
	//           at this point each column may have >= 1 field name options;
	//           later, we will prune this to a discrete solution.
	optsPerColumn := []map[string]bool{}
	numCols := len(tickets[0].values)
	for c := 0; c < numCols; c += 1 {
		colOpts := []string{}
		for _, candField := range listFieldNames(fields) {
			constraint := fields[candField]
			validForAll := true
			for _, ticket := range tickets {
				if !constraint.validate(ticket.col(c)) {
					if c == 2 {
					}
					validForAll = false
				}
			}
			if validForAll {
				colOpts = append(colOpts, candField)
			}
		}
		if c == 2 {
		}
		optsPerColumn = append(optsPerColumn, lst2set(colOpts))
	}

	// Part 2 :: solve the field constraints.
	//           just figure out what column has to be a certain field, then
	//           reduce all other column options; rinse and repeat until
	//           we're done.
	allDone := func(solution []map[string]bool) bool {
		for _, opts := range solution {
			if len(opts) > 1 {
				return false
			}
		}
		return true
	}
	solved := map[int]bool{}
	for !allDone(optsPerColumn) {
	findColSolution:
		for n, opts := range optsPerColumn {
			if len(opts) == 1 && !solved[n] {
				constrainedSolution := singleton(opts)
				for m, otherOpts := range optsPerColumn {
					if n == m {
						continue
					}
					delete(otherOpts, constrainedSolution)
				}
				solved[n] = true
				break findColSolution
			}
		}
	}

	order := []string{}
	for _, field := range optsPerColumn {
		order = append(order, singleton(field))
	}
	return order
}

func productFieldsDeparture(fields []string, t ticket) int {
	product := 1
	for i, field := range fields {
		if strings.HasPrefix(field, "departure") {
			product *= t.col(i)
		}
	}
	return product
}

func main() {
	fi, err := ioutil.ReadFile("day16/input.txt")
	check(err)
	ficontent := string(fi)

	fields, yourTicket, nearbyTickets := parse(ficontent).values()

	fmt.Println("Part 1:")
	fmt.Println(nearbyTicketErrorRate(fields, nearbyTickets))

	fmt.Println("Part 2:")
	validTickets := append(filterValid(fields, nearbyTickets), yourTicket)
	fieldOrder := solveFieldOrder(fields, validTickets)
	departureProduct := productFieldsDeparture(fieldOrder, yourTicket)
	fmt.Println(departureProduct)
}
